python/device/running_operation_list.py

```diff
@@ -0,0 +1,133 @@
+        """
+        remove an operation fron the list which owns it.
+        """
+        owner = self.owner_weakref and self.owner_weakref()
+        if owner:
+            if in_dunder_del:
+                logger.warning("Abandoning an incompleted operation: id={}".format(self.id))
+            owner.remove(self.id)
+            self.owner_weakref = None
+
+
+class EventBasedOperation(OperationBase):
+    """
+    Running operation which sets an event when it is complete.
+    """
+
+    def __init__(self, owner):
+        self.owner_weakref = weakref.ref(owner)
+        self.id = str(uuid.uuid4())
+        self.event = threading.Event()
+
+    def complete(self):
+        """
+        Mark a running operation as complete.  This sets the operation's event and removes
+        the operation from the list which owns it.
+        """
+        if self.event:
+            self.event.set()
+            self.event = None
+            self.remove_from_owning_list()
+        else:
+            logger.warning("Attempt to complete already completed operation: id={}".format(self.id))
+
+
+class CallbackBasedOperation(OperationBase):
+    """
+    Running operation which calls a callback when it is complete.
+    """
+
+    def __init__(self, owner, callback, user_data):
+        self.owner_weakref = weakref.ref(owner)
+        self.id = str(uuid.uuid4())
+        self.callback = callback
+        self.user_data = user_data
+
+    def complete(self):
+        """
+        Mark a running operation as complete.  This calls the operation's callback and removes
+        the operation from the list which owns it.
+        """
+        if self.callback:
+            self.callback(self.id, self.user_data)
+            self.callback = None
+            self.remove_from_owning_list()
+        else:
+            logger.warning("Attempt to complete already completed operation: id={}".format(self.id))
+
+
+class RunningOperationList(object):
+    """
+    Object which keeps tracks of operations which are running (in progress, not yet complete.)
+
+    The "running operation" objects in this list have two properties which make them useful:
+    1. They have an automatically-generated ID value (a guid) which can be passed as a serviceAckId
+    2. They have a complete method, which can either set a `threading.Event` object or call a callback.
+
+    Using these objects, we can have a guid (a `serviceAckId`) which is like a "completion token".
+    When the server returns the serviceAckId, this list can use that guid to run some "on complete"
+    code, which might be a callback, or might be a `threading.Event` object.
+    """
+
+    def __init__(self):
+        self.lock = threading.Lock()
+        self.list = {}
+
+    def make_event_based_operation(self):
+        """
+        Make and return a running opreation object which fires an event when it is complete.
+        """
+        e = EventBasedOperation(self)
```

Comment:
nit: Does `e` stand for `event` here? I'd rename this to something that describes it's purpose better. Same feedback applies elsewhere in the code.